internal/core/model.rs

```diff
@@ -1226,3 +1249,253 @@ fn test_vecmodel_set_vec() {
+                .inner_model
+                .iter()
+                .enumerate()
+                .filter(|(_, e)| (self.filter_function)(e))
+                .map(|(i, _)| i)
+                .collect();
+        }
+    }
+
+    impl<M, F, T> Model for FilterProxy<M, F, T>
+    where
+        M: Model<Data = T> + 'static,
+        F: Fn(&T) -> bool + 'static,
+        T: 'static,
+    {
+        type Data = T;
+
+        fn row_count(&self) -> usize {
+            self.mapping.borrow().len()
+        }
+
+        fn row_data(&self, row: usize) -> Option<Self::Data> {
+            self.mapping
+                .borrow()
+                .get(row)
+                .map(|&inner_row| self.inner_model.row_data(inner_row).unwrap())
+        }
+
+        fn model_tracker(&self) -> &dyn ModelTracker {
+            &self.notify
+        }
+    }
+
+    impl<M, F, T> ErasedRepeater for FilterProxy<M, F, T>
+    where
+        M: Model<Data = T> + 'static,
+        F: Fn(&T) -> bool + 'static,
+        T: 'static,
+    {
+        fn row_changed(&self, row: usize) {
+            let mut mapping = self.mapping.borrow_mut();
+
+            let index;
+            let is_contained = match mapping.binary_search(&row) {
+                Ok(i) => {
+                    index = i;
+                    true
+                }
+                Err(i) => {
+                    index = i;
+                    false
+                }
+            };
+
+            let should_be_contained =
+                (self.filter_function)(&self.inner_model.row_data(row).unwrap());
+
+            if is_contained && should_be_contained {
+                drop(mapping);
+                self.notify.row_changed(index);
+            } else if !is_contained && should_be_contained {
+                mapping.insert(index, row);
+                drop(mapping);
+                self.notify.row_added(index, 1);
+            } else if is_contained && !should_be_contained {
+                mapping.remove(index);
+                drop(mapping);
+                self.notify.row_removed(index, 1);
+            }
+        }
+
+        fn row_added(&self, index: usize, count: usize) {
+            if count == 0 {
+                return;
+            }
+
+            let insertion: Vec<usize> = self
+                .inner_model
+                .iter()
+                .enumerate()
+                .skip(index)
+                .take(count)
+                .filter_map(|(i, e)| if (self.filter_function)(&e) { Some(i) } else { None })
+                .collect();
+
+            if insertion.len() > 0 {
+                let mut mapping = self.mapping.borrow_mut();
+                let insertion_point = match mapping.binary_search(&index) {
+                    Ok(ip) => ip,
+                    Err(ip) => ip,
+                };
+
+                for (value, &count) in insertion.iter().enumerate() {
+                    mapping.insert(insertion_point + count, value);
+                }
+
```

Comment:
Also remember that all the values after the insertion_point need to be incremented by count.